src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java

```diff
@@ -226,19 +240,32 @@ public void periodic() {
                 } else {
                         updateRotationTarget = true;
                 }
+
+                if(Robot.isSimulation()) {
+                        if(!DriverStationSim.getAutonomous()) {
+                                states = kinematics.toSwerveModuleStates(chassisSpeeds);
+                        }
+                        positionMeters[0] += states[0].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[1] += states[1].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[2] += states[2].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[3] += states[3].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        simRotation += chassisSpeeds.omegaRadiansPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        
+                }else {
+                        positionMeters[0] += frontLeftModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime); // (m / s) * delta t = m
+                        positionMeters[1] += frontRightModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[2] += backLeftModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[3] += backRightModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime);
+                }
```

Comment:
I think this if statement needs to be moved below the module setting as there is still some code below that can modify the state ```desaturate``` therefore it would be possible we are using the non-final states for the simulation.

In face I would recommend wrapping this if statement + the field.setRobotPose() call in a new method that is called ```updateOdometry()```. Then call that method at the end of ```periodic```

Reply:
I will move forward and approve but just wanted to comment that I think the full above if statement is only related to the updating of the Pose so it ideally could live inside of updatePose as well. Something to consider in future if periodic() grows in complexity again.